### PR TITLE
Client mapblocks: do frustum culling via planes

### PR DESCRIPTION
**Pull request by Desour**
- Goal of the PR: Performance. Also, the new method is simpler.
- How does the PR work?: Use the planes of the frustum. Each such plane gives a half-space. Check whether the mapblock is completely outside of one of this half-spaces by using the radius of the mapblock mesh bounding sphere and the signed distance of this sphere's center to the plane. This culls more mapblocks (about 1/6).
- Does this relate to a goal in [the roadmap](../doc/direction.md)? Yes. It's graphics performance improvements, (and internal code refactoring).
- This is also a bugfix: Very long plantlike rooted nodes are no longer culled too early.

## To do

This PR is a Ready for Review.

Please squash on merge.

## How to test

* Revert that one last revert commit (to add back the quicktune code stuff).
* In game, use the quicktune keys (pos1, page up, page down, ... on numpad) to use old or new culling.
* Look at fps and drawtime. (New is around 0 to 5 fps faster for me, depending on scene.)
* Also move your camera around and look for not drawn mapblocks.

### CODE REVIEW

File: src/client/camera.cpp [hunk 1]
```diff
@@ -463,6 +467,7 @@ void Camera::update(LocalPlayer* player, f32 frametime, f32 tool_reload_ratio)
 
 	// Set camera node transformation
 	m_cameranode->setPosition(my_cp-intToFloat(m_camera_offset, BS));
+	m_cameranode->updateAbsolutePosition();
```
Comment (Desour) [inline src/client/camera.cpp:470]: See <https://github.com/minetest/irrlicht/blob/ff645cc8764ab3fc79f0e67677a328ba36c35087/include/ICameraSceneNode.h#L83>.

File: src/client/camera.cpp [hunk 2]
```diff
@@ -706,3 +715,15 @@ void Camera::removeNametag(Nametag *nametag)
 	m_nametags.remove(nametag);
 	delete nametag;
 }
+
+std::array<core::plane3d<f32>, 4> Camera::getFrustumCullPlanes() const
```
Comment (Desour) [inline src/client/camera.cpp:719]: This function exists as a compile time optimisation (to not include `<SViewFrustum.h>` in the header).

File: src/client/camera.h [hunk 3]
```diff
@@ -133,6 +135,22 @@ class Camera
 		return MYMAX(m_fov_x, m_fov_y);
 	}
 
+	// Returns a lambda that when called with an object's position (in BS space)
+	// returns true if, and only if the object should be frustum-culled.
+	// radius is the radius of the object.
+	auto getFrustumCuller(f32 radius) const
```
Comment (Desour) [inline src/client/camera.h:141]: (Might be useful in a later PR for frustum culling other objects, like CAOs, btw..)

File: src/client/clientmap.cpp [hunk 4]
```diff
@@ -271,22 +267,19 @@ void ClientMap::updateDrawList()
...
 
 			// Frustum culling
-			float d = 0.0;
-			if (!isBlockInSight(block_coord, camera_position,
-					camera_direction, camera_fov, range * BS, &d))
+			if (is_frustum_culled(intToFloat(block_position, BS)))
 				continue;
 
 			// Occlusion culling
-			if ((!m_control.range_all && d > m_control.wanted_range * BS) ||
```
Comment (Desour) [inline src/client/clientmap.cpp:288]: Distance check is already done above.

**Comment by lhofhansl** [comment]
The server (clientiface.cpp) also uses isBlockInSight. Should we optimize it there as well?

**Comment by Desour** [comment]
> The server (clientiface.cpp) also uses isBlockInSight. Should we optimize it there as well?

I didn't want to cause conflicts with #12690. So the answer is: currently not.
Also, more precise culling on the server-side probably doesn't give too much profit.

**Comment by Desour** [comment]
I've added code for calculating the actual bounding-box of the mapblock mesh. Most of the times it's smaller than `BLOCK_MAX_RADIUS` (=> cull more), but there are also drawtypes that need this, i.e. test with the devtest rooted plantlike leveled node.
(Btw., FYI, `MapblockMeshGenerator` seems to generate vertex positions from `v3f(-0.5f*BS)` to `v3f(15.5f*BS)` (for normal drawtype).)

File: src/client/camera.cpp (hunk 2 above)
Comment (SmallJoker) [reply src/client/camera.cpp:719]: In that case it should be noted down as a comment so that it's not "cleaned up" in a future change

**Review by SmallJoker — APPROVED** [review]
Viewing range: 160
Scene: generated MTG world, few trees, mountains, river.

PR: 215 FPS
master: 165 FPS

EDIT: See also https://github.com/minetest/minetest/pull/12710#discussion_r967687248

File: src/client/camera.cpp (hunk 2 above)
Comment (Desour) [reply src/client/camera.cpp:719]: Done.
(Though this detail is not really important. I just noted this here to explain why I added the function. If someone would do a clean up, they'd notice that they have to include more in the header. Anyway, a comment doesn't hurt. `¯\_(ツ)_/¯`)

**Comment by lhofhansl** [comment]
Seems to work fine. In my test I see indeed a jump from 32 fps to 38 fps.
I also see the number of drawn blocks drop from 721 to 680, which I found a bit surprising.
Perhaps because the plane-culling is more accurate?

File: src/client/camera.cpp (hunk 1 above)
Comment (lhofhansl) [reply src/client/camera.cpp:470]: Good. But unrelated, right?

**Comment by lhofhansl** [comment]
Did not see any missing blocks, and code look reasonable :)

**Pull request merged**